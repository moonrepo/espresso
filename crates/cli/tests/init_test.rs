mod utils;

use espresso_manifest::MANIFEST_NAME;
use starbase_sandbox::predicates::prelude::*;
use starbase_sandbox::{assert_snapshot, create_empty_sandbox};
use utils::*;

mod init_polyrepo {
    use espresso_lockfile::LOCKFILE_NAME;

    use super::*;

    #[test]
    fn errors_if_name_not_provided() {
        let sandbox = create_empty_sandbox();

        create_espm_command(sandbox.path())
            .args(["init", "--yes"])
            .env("ESPM_INIT_WORKSPACE", "0")
            .assert()
            .failure()
            .stderr(predicate::str::contains(
                "A package name is required with --name when using --yes.",
            ));
    }

    #[test]
    fn creates_package() {
        let sandbox = create_empty_sandbox();

        create_espm_command(sandbox.path())
            .args(["init", "--yes", "--name", "namespace/package"])
            .env("ESPM_INIT_WORKSPACE", "0")
            .assert()
            .success();

        assert!(!sandbox.path().join(LOCKFILE_NAME).exists());
        assert!(sandbox.path().join(MANIFEST_NAME).exists());
        assert!(sandbox.path().join("README.md").exists());
        assert!(sandbox.path().join("src/index.ts").exists());
    }
}

mod init_monorepo {
    use espresso_lockfile::LOCKFILE_NAME;

    use super::*;

    #[test]
    fn errors_if_to_already_a_package() {
        let sandbox = create_empty_sandbox();
        sandbox.create_file(MANIFEST_NAME, "");

        create_espm_command(sandbox.path())
            .args(["init", "--yes", "--to", "."])
            .env("ESPM_INIT_WORKSPACE", "1")
            .assert()
            .failure()
            .stderr(predicate::str::contains(
                "A package or workspace already exists",
            ));
    }

    #[test]
    fn creates_workspace() {
        let sandbox = create_empty_sandbox();

        create_espm_command(sandbox.path())
            .args(["init", "--yes"])
            .env("ESPM_INIT_WORKSPACE", "1")
            .assert()
            .success();

        assert!(sandbox.path().join(LOCKFILE_NAME).exists());
        assert!(sandbox.path().join(MANIFEST_NAME).exists());
        assert!(sandbox.path().join("packages").exists());
        assert!(!sandbox.path().join("src/index.ts").exists());

        assert_snapshot!(read_file(sandbox.path().join(MANIFEST_NAME)));
    }
}
