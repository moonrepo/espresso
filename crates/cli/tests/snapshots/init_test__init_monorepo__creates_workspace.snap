---
source: crates/cli/tests/init_test.rs
expression: read_file(sandbox.path().join(MANIFEST_NAME))
---
[workspace]
packages = ["packages/*"]

